pkg/trtl/replica/sync.go

```diff
@@ -0,0 +1,491 @@
+// routines, so 4 go routines are operating on the initiator side to handle the sync.
+// The first phase go routine ends when it finishes looping over its database, the
+// second phase go routine is also the recv go routine so it starts shortly after the
+// first phase go routine and runs concurrently with it. The second phase ends when it
+// receives complete from the remote. The send go routine ends when there are no more
+// messages on its channel. Once all go routines are completed the initiator closes the
+// channel, ending the synchronization between the initiator and the remote.
+func (r *Service) AntiEntropySync(peer *peers.Peer, log zerolog.Logger) (err error) {
+	// Create a context with a timeout that is sooner than 95% of the timeouts selected
+	// by the normally distributed jittered interval, to ensure anti-entropy gossip
+	// sessions do not span multiple anti-entropy intervals.
+	timeout := r.conf.GossipInterval - (2 * r.conf.GossipSigma)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Dial the remote peer and establish a connection
+	var cc *grpc.ClientConn
+	if cc, err = r.connect(ctx, peer); err != nil {
+		return err
+	}
+	defer cc.Close()
+
+	// Initiate the Gossip Stream
+	client := replica.NewReplicationClient(cc)
+	var stream replica.Replication_GossipClient
+	if stream, err = client.Gossip(ctx); err != nil {
+		return fmt.Errorf("could not connect gossip stream: %s", err)
+	}
+
+	// Report successful connection
+	log.Debug().Msg("dialed remote peer and connected to the gossip stream")
+
+	// Create a stream sender to ensure that both phase1 and phase2 go routines can send
+	// messages concurrently without violating the grpc semantic that only one go
+	// routine can send messages on the stream.
+	// NOTE: newStreamSender calls wg.Add(1) and runs the sender go routine.
+	wg := new(sync.WaitGroup)
+	sender := newStreamSender(wg, log, stream)
+
+	// Start phase 1: loop over all objects in the local database and send check
+	// requests to the remote replica. This is also called the "pull" phase, since we're
+	// asking the remote for its objects that are later than our own, e.g. pulling the
+	// objects to this replica from the remote. Phase 1 ends when we've completed
+	// looping over the local database.
+	wg.Add(1)
+	go r.initiatorPhase1(ctx, wg, log, sender)
+
+	// Start phase 2: this phase is concurrent with phase 1 since it listens for and
+	// responds to all messages from the remote replica. This is also called the "push"
+	// phase, since we're pushing objects back to the remote replica. This phase ends
+	// when the remote replica sends a COMPLETE message, meaning it is done sending
+	// messages. At that point, we will no longer send any messages so this phase will
+	// close the sender go routine, which will stop when all messages have been sent.
+	wg.Add(1)
+	go r.initiatorPhase2(ctx, wg, log, sender, stream)
```

Comment:
Interesting! Possibly I had this backwards before. I was under the impression that phase one was called the push phase because the initiator is pushing (first versions, then data) to the remote. Push is all that happens in the non bilateral AE case, and the remote is the only replica that makes changes in that case. Bilateral AE introduces phase two, the pull phase, where the initiator has the opportunity to receive repairs from the remote. 

If I do have that backwards, please let me know and I'll fix the [anti-entropy docs](https://docs.google.com/presentation/d/1HhP5PwV5vyqU93SQsLx3IwD_5t-vPZFxYF8kTIiMfEc/edit?usp=sharing) I wrote.

Reply:
I think you have it right. Just to clarify; I believe that push and pull refer to types of gossip where the synchronization is unidirectional:

1. Pull: initiator is synchronized with remote
2. Push: remote is synchronized by the initiator 
3. Bilateral: both initiator and remote are synchronized